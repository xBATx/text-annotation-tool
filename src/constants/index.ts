import {
  Annotation as AnnotationType,
  Relation as RelationType,
  AnnotationCategoryType,
  RelationCategoryType,
} from 'src/types';

export const SELECT_TEXT = 'SELECT_TEXT';

export const REMOVE_SELECTED_ELEMENT = 'REMOVE_SELECTED_ELEMENT';
export const RESET_SELECTION = 'RESET_SELECTION';

export const ADD_ANNOTATION = 'ADD_ANNOTATION';
export const SELECT_ANNOTATION = 'SELECT_ANNOTATION';

export const PREPARE_RELATION = 'PREPARE_RELATION';
export const ADD_RELATION = 'ADD_RELATION';
export const SELECT_RELATION = 'SELECT_RELATION';
export const SET_SHOW_RELATIONS = 'SET_SHOW_RELATIONS';

// Modal actions
export const SUBMIT_ARTICLE_MODAL = 'SUBMIT_ARTICLE_MODAL';
export const SET_ARTICLE_MODAL_OPENED = 'SET_ARTICLE_MODAL_OPENED';
export const SET_ARTICLE_MODAL_TEXT = 'SET_ARTICLE_MODAL_TEXT';

// A8ns category types
export const FIRST_NAME = 'FIRST_NAME';
export const LAST_NAME = 'LAST_NAME';
export const FULL_NAME = 'FULL_NAME';
export const COMPANY_NAME = 'COMPANY_NAME';
export const DOB = 'DOB';
export const RISK = 'RISK';

// Relation types between a8ns
export const COREFERENCE = 'COREFERENCE';
export const INVOLVES = 'INVOLVES';

export const AllA8nCategories = [
  FIRST_NAME,
  LAST_NAME,
  FULL_NAME,
  COMPANY_NAME,
  DOB,
  RISK,
];
export const AllRelationCategories = [COREFERENCE, INVOLVES];

export const CategoryColorMap: { [K in AnnotationCategoryType]: string } = {
  FIRST_NAME: 'lightblue',
  LAST_NAME: 'lightgreen',
  FULL_NAME: 'orange',
  COMPANY_NAME: 'cyan',
  DOB: 'violet',
  RISK: 'red',
};

export const RelationColorMap: { [K in RelationCategoryType]: string } = {
  COREFERENCE: 'purple',
  INVOLVES: 'blue',
};

export const SelectedAnnotationColor = 'yellow';
export const SelectedRelationColor = 'red';

export const SampleAnnotation: AnnotationType = {
  id: '0:1:FULL_NAME',
  renderingId: '0:1:FULL_NAME',
  from: 0,
  to: 1,
  category: FULL_NAME,
  text: 'L',
};

export const SampleRelation: RelationType = {
  id: 'A->B',
  from: 'A',
  to: 'B',
  category: COREFERENCE,
};

export const SampleText =
  'Lorem ipsum dolor sit amet, nominavi rationibus vix in. Pro' +
  'omittam intellegam ea. Eum malorum nonumes oporteat cu, qui illum' +
  'imperdiet id, deserunt salutandi theophrastus ut quo. Quo summo' +
  'minimum scribentur an, no mel sumo oratio definitiones. Vim at' +
  'utamur luptatum lobortis, urbanitas vituperata disputationi eu' +
  'vis, eos ut iuvaret mentitum signiferumque. Vix tota convenire no,' +
  'meis mandamus facilisis cum an. Velit patrioque complectitur est' +
  'ex. Id vide debitis suscipiantur quo. Virtute adolescens eam eu.' +
  'Id phaedrum posidonium deterruisset quo. Mei et justo tempor. Duo' +
  'oratio labores sensibus ea, ad modo quas graece qui. Cu ancillae' +
  'explicari nam, viris numquam cu vim. Id amet consequuntur vis, mei' +
  'oporteat legendos oportere in, in sit labores perfecto' +
  'adversarium. Mucius voluptatum sea in. His et movet delectus. Sint' +
  'duis brute ne nec. Nam ex purto mazim volutpat, enim philosophia' +
  'ut duo. Vim ad justo dicta paulo. Mel posse discere eu, placerat' +
  'comprehensam et vel. Sed et assum disputationi, utroque accusamus' +
  'laboramus ex cum, vel populo tractatos ex. Putent pericula te vix,' +
  'nam an lobortis dissentias. At veniam dicunt concludaturque est,' +
  'impetus reprimique efficiendi vel ne. In ius putent aperiri' +
  'vivendum, an nam amet graecis sadipscing, summo nominati et sed.' +
  'Facete doctus ex sit. Vim malorum persequeris no.' +
  '\n\n' +
  'Lorem ipsum dolor sit amet,' +
  'rationibus vix in. Pro omittam intellegam ea. Eum malorum nonumes' +
  'oporteat cu, qui illum imperdiet id, deserunt salutandi' +
  'theophrastus ut quo. Quo summo minimum scribentur an, no mel sumo' +
  'oratio definitiones. Vim at utamur luptatum lobortis, urbanitas' +
  'vituperata disputationi eu vis, eos ut iuvaret mentitum' +
  'signiferumque. Vix tota convenire no, meis mandamus facilisis cum' +
  'an. Velit patrioque complectitur est ex. Id vide debitis' +
  'suscipiantur quo. Virtute adolescens eam eu. Id phaedrum' +
  'posidonium deterruisset quo. Mei et justo tempor. Duo oratio' +
  'labores sensibus ea, ad modo quas graece qui. Cu ancillae' +
  'explicari nam, viris numquam cu vim. Id amet consequuntur vis, mei' +
  'oporteat legendos oportere in, in sit labores perfecto' +
  'adversarium. Mucius voluptatum sea in. His et movet delectus. Sint' +
  'duis brute ne nec. Nam ex purto mazim volutpat, enim philosophia' +
  'ut duo. Vim ad justo dicta paulo. Mel posse discere eu, placerat' +
  'comprehensam et vel. Sed et assum disputationi, utroque accusamus' +
  'laboramus ex cum, vel populo tractatos ex. Putent pericula te vix,' +
  'nam an lobortis dissentias. At veniam dicunt concludaturque est,' +
  'impetus reprimique efficiendi vel ne. In ius putent aperiri' +
  'vivendum, an nam amet graecis sadipscing, summo nominati et sed.' +
  'Facete doctus ex sit. Vim malorum persequeris no.' +
  '\n\n' +
  'Lorem ipsum dolor sit amet, nominavi rationibus vix in. Pro' +
  'omittam intellegam ea. Eum malorum nonumes oporteat cu, qui illum' +
  'imperdiet id, deserunt salutandi theophrastus ut quo. Quo summo' +
  'minimum scribentur an, no mel sumo oratio definitiones. Vim at' +
  'utamur luptatum lobortis, urbanitas vituperata disputationi eu' +
  'vis, eos ut iuvaret mentitum signiferumque. Vix tota convenire no,' +
  'meis mandamus facilisis cum an. Velit patrioque complectitur est' +
  'ex. Id vide debitis suscipiantur quo. Virtute adolescens eam eu.' +
  'Id phaedrum posidonium deterruisset quo. Mei et justo tempor. Duo' +
  'oratio labores sensibus ea, ad modo quas graece qui. Cu ancillae' +
  'explicari nam, viris numquam cu vim. Id amet consequuntur vis, mei' +
  'oporteat legendos oportere in, in sit labores perfecto' +
  'adversarium. Mucius voluptatum sea in. His et movet delectus. Sint' +
  'duis brute ne nec. Nam ex purto mazim volutpat, enim philosophia' +
  'ut duo. Vim ad justo dicta paulo. Mel posse discere eu, placerat' +
  'comprehensam et vel. Sed et assum disputationi, utroque accusamus' +
  'laboramus ex cum, vel populo tractatos ex. Putent pericula te vix,' +
  'nam an lobortis dissentias. At veniam dicunt concludaturque est,' +
  'impetus reprimique efficiendi vel ne. In ius putent aperiri' +
  'vivendum, an nam amet graecis sadipscing, summo nominati et sed.' +
  'Lorem ipsum dolor sit amet,' +
  'rationibus vix in. Pro omittam intellegam ea. Eum malorum nonumes' +
  'oporteat cu, qui illum imperdiet id, deserunt salutandi' +
  'theophrastus ut quo. Quo summo minimum scribentur an, no mel sumo' +
  'oratio definitiones. Vim at utamur luptatum lobortis, urbanitas' +
  'vituperata disputationi eu vis, eos ut iuvaret mentitum' +
  'signiferumque. Vix tota convenire no, meis mandamus facilisis cum' +
  'an. Velit patrioque complectitur est ex. Id vide debitis' +
  'suscipiantur quo. Virtute adolescens eam eu. Id phaedrum' +
  'posidonium deterruisset quo. Mei et justo tempor. Duo oratio' +
  'labores sensibus ea, ad modo quas graece qui. Cu ancillae' +
  'explicari nam, viris numquam cu vim. Id amet consequuntur vis, mei' +
  'oporteat legendos oportere in, in sit labores perfecto' +
  'adversarium. Mucius voluptatum sea in. His et movet delectus. Sint' +
  'duis brute ne nec. Nam ex purto mazim volutpat, enim philosophia' +
  'ut duo. Vim ad justo dicta paulo. Mel posse discere eu, placerat' +
  'comprehensam et vel. Sed et assum disputationi, utroque accusamus' +
  'laboramus ex cum, vel populo tractatos ex. Putent pericula te vix,' +
  'nam an lobortis dissentias. At veniam dicunt concludaturque est,' +
  'impetus reprimique efficiendi vel ne. In ius putent aperiri' +
  'vivendum, an nam amet graecis sadipscing, summo nominati et sed.' +
  'Facete doctus ex sit. Vim malorum persequeris no.' +
  '\n\n' +
  'Lorem ipsum dolor sit amet, nominavi rationibus vix in. Pro' +
  'omittam intellegam ea. Eum malorum nonumes oporteat cu, qui illum' +
  'imperdiet id, deserunt salutandi theophrastus ut quo. Quo summo' +
  'minimum scribentur an, no mel sumo oratio definitiones. Vim at' +
  'utamur luptatum lobortis, urbanitas vituperata disputationi eu' +
  'vis, eos ut iuvaret mentitum signiferumque. Vix tota convenire no,' +
  'meis mandamus facilisis cum an. Velit patrioque complectitur est' +
  'ex. Id vide debitis suscipiantur quo. Virtute adolescens eam eu.' +
  'Id phaedrum posidonium deterruisset quo. Mei et justo tempor. Duo' +
  'oratio labores sensibus ea, ad modo quas graece qui. Cu ancillae' +
  'explicari nam, viris numquam cu vim. Id amet consequuntur vis, mei' +
  'oporteat legendos oportere in, in sit labores perfecto' +
  'adversarium. Mucius voluptatum sea in. His et movet delectus. Sint' +
  'duis brute ne nec. Nam ex purto mazim volutpat, enim philosophia' +
  'ut duo. Vim ad justo dicta paulo. Mel posse discere eu, placerat' +
  'comprehensam et vel. Sed et assum disputationi, utroque accusamus' +
  'laboramus ex cum, vel populo tractatos ex. Putent pericula te vix,' +
  'nam an lobortis dissentias. At veniam dicunt concludaturque est,' +
  'impetus reprimique efficiendi vel ne. In ius putent aperiri' +
  'vivendum, an nam amet graecis sadipscing, summo nominati et sed.' +
  'Facete doctus ex sit. Vim malorum persequeris no.';
