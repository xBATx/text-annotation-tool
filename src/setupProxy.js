// @ts-ignore
const proxy = require("http-proxy-middleware");

module.exports = function(app) {
    app.use(proxy("/api", {
        "pathRewrite": {
            "^/api": "",
        },
        "target": "http://35.237.122.46:8000",
    }));
};
