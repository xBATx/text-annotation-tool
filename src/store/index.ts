import { createStore, compose } from 'redux';
import rootReducer from 'src/reducers';

// declare global {
//   interface Window {
//     __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
//   }
// }

// @ts-ignore
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default createStore(rootReducer, undefined, composeEnhancers());
