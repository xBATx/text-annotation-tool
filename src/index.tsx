import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import Annotation from './components/annotation';
import store from './store';

import './index.css';

const App = () => (
  <Provider store={store}>
    <Annotation />
  </Provider>
);

ReactDOM.render(<App />, document.getElementById('root'));
