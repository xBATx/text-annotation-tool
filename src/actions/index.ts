import { Action } from 'redux';
import {
  SUBMIT_ARTICLE_MODAL,
  ADD_ANNOTATION,
  SELECT_TEXT,
  REMOVE_SELECTED_ELEMENT,
  RESET_SELECTION,
  SELECT_ANNOTATION,
  SET_ARTICLE_MODAL_OPENED,
  SET_ARTICLE_MODAL_TEXT,
  ADD_RELATION,
  SELECT_RELATION,
  SET_SHOW_RELATIONS,
  PREPARE_RELATION,
} from 'src/constants';
import { Annotation, Relation, SelectedText } from 'src/types';

// Action types

interface SelectTextType extends Action {
  type: typeof SELECT_TEXT;
  selectedText: SelectedText | null;
}

interface RemoveSelectedElementType extends Action {
  type: typeof REMOVE_SELECTED_ELEMENT;
}
interface ResetSelectionType extends Action {
  type: typeof RESET_SELECTION;
}

interface AddAnnotationType extends Action {
  type: typeof ADD_ANNOTATION;
  annotation: Annotation;
}
interface SelectAnnotationType extends Action {
  type: typeof SELECT_ANNOTATION;
  id: string | null;
}

interface PrepareRelationType extends Action {
  type: typeof PREPARE_RELATION;
  toA8nId: string | null;
}
interface AddRelationType extends Action {
  type: typeof ADD_RELATION;
  relation: Relation;
}
interface SelectRelationType extends Action {
  type: typeof SELECT_RELATION;
  id: string | null;
}
interface SetShowRelationsType extends Action {
  type: typeof SET_SHOW_RELATIONS;
  show: boolean;
}

// Article modal action types
interface SetArticleModalOpenedType extends Action {
  type: typeof SET_ARTICLE_MODAL_OPENED;
  opened: boolean;
}
interface SetArticleModalTextType extends Action {
  type: typeof SET_ARTICLE_MODAL_TEXT;
  text: string;
}
interface SubmitArticleModalType extends Action {
  type: typeof SUBMIT_ARTICLE_MODAL;
  text: string;
  annotations: Annotation[];
  relations: Relation[];
}

// Action factories

export const SelectText = (
  selectedText: SelectedText | null,
): SelectTextType => ({
  type: SELECT_TEXT,
  selectedText,
});
export const RemoveSelectedElement: RemoveSelectedElementType = {
  type: REMOVE_SELECTED_ELEMENT,
};
export const ResetSelection: ResetSelectionType = {
  type: RESET_SELECTION,
};

export const AddAnnotation = (annotation: Annotation): AddAnnotationType => ({
  type: ADD_ANNOTATION,
  annotation,
});
export const SelectAnnotation = (id: string | null): SelectAnnotationType => ({
  type: SELECT_ANNOTATION,
  id,
});

export const PrepareRelation = (
  toA8nId: string | null,
): PrepareRelationType => ({
  type: PREPARE_RELATION,
  toA8nId,
});
export const AddRelation = (relation: Relation): AddRelationType => ({
  type: ADD_RELATION,
  relation,
});
export const SelectRelation = (id: string | null): SelectRelationType => ({
  type: SELECT_RELATION,
  id,
});
export const SetShowRelations = (show: boolean): SetShowRelationsType => ({
  type: SET_SHOW_RELATIONS,
  show,
});

// Add article modal actions
export const SetArticleModalOpened = (
  opened: boolean,
): SetArticleModalOpenedType => ({
  type: SET_ARTICLE_MODAL_OPENED,
  opened,
});
export const SetArticleInput = (text: string): SetArticleModalTextType => ({
  type: SET_ARTICLE_MODAL_TEXT,
  text,
});
export const SubmitArticleModal = (
  text: string,
  annotations: Annotation[],
  relations: Relation[],
): SubmitArticleModalType => ({
  type: SUBMIT_ARTICLE_MODAL,
  text,
  annotations,
  relations,
});

export type AppActions =
  | SelectTextType
  | RemoveSelectedElementType
  | ResetSelectionType
  | AddAnnotationType
  | SelectAnnotationType
  | PrepareRelationType
  | AddRelationType
  | SelectRelationType
  | SetShowRelationsType
  | SetArticleModalOpenedType
  | SetArticleModalTextType
  | SubmitArticleModalType;
