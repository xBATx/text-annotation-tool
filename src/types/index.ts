import { AllRelationCategories, AllA8nCategories } from 'src/constants';

export type AnnotationCategoryType = typeof AllA8nCategories[number];

export type RelationCategoryType = typeof AllRelationCategories[number];

export interface SelectedText {
  from: number;
  to: number;
  text: string;
}

export interface Annotation {
  id: string;
  renderingId: string;
  from: number;
  to: number;
  category: AnnotationCategoryType;
  text: string;
}

export interface Relation {
  id: string;
  from: string;
  to: string;
  category: RelationCategoryType;
}

export interface ArticleModalState {
  isOpened: boolean;
  inputText: string;
}

export interface AnnotationState {
  text: string;
  annotations: Annotation[];
  relations: Relation[];
  showRelations: boolean;
  selectedText: SelectedText | null;
  selectedAnnotation: string | null;
  prepareRelationToA8nId: string | null;
  selectedRelation: Relation | null;
  articleModalState: ArticleModalState;
}

export interface AppState {
  annotation: AnnotationState;
}
