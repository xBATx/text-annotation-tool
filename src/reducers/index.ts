import { Action } from 'redux';
import annotation, {
  initialState as annotationInitialState,
} from './annotation';
import { AppState } from 'src/types';
import { AppActions } from 'src/actions';

const defaultState = {
  annotation: annotationInitialState,
};

const rootReducer = (state: AppState = defaultState, action: Action) => ({
  annotation: annotation(state.annotation, action as AppActions),
});

export default rootReducer;
