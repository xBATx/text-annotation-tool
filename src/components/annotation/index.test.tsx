import React from 'react';
import renderer from 'react-test-renderer';
import { Annotation } from './index';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { AnnotationState } from 'src/types';

const annotationState: AnnotationState = {
  text: 'Lorem ipsum dolor sit amet, nominavi rationibus vix in.',
  annotations: [],
  relations: [],
  showRelations: true,
  articleModalState: { isOpened: false, inputText: '' },
  selectedText: null,
  selectedRelation: null,
  selectedAnnotation: null,
  prepareRelationToA8nId: null,
};

it('renders without crashing', () => {
  const mockStore = configureStore();
  const showRelations = true;
  const setShowRelations = jest.fn();
  const wrapper = renderer
    .create(
      <Provider store={mockStore({ annotation: annotationState })}>
        <Annotation
          showRelations={showRelations}
          setShowRelations={setShowRelations}
        />
      </Provider>,
    )
    .toJSON();
  expect(wrapper).toMatchSnapshot();
});
