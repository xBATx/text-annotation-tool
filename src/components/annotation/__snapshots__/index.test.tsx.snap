// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
Array [
  <div
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "blue",
        "display": "flex",
        "height": 60,
        "justifyContent": "flex-start",
        "position": "relative",
        "width": "100%",
        "zIndex": 4,
      }
    }
  >
    <a
      href="https://github.com/xBATx/text-annotation-tool"
      style={
        Object {
          "left": "15px",
          "position": "absolute",
        }
      }
    >
      <img
        src="github.png"
        style={
          Object {
            "color": "white",
            "cursor": "pointer",
            "height": "25px",
            "width": "25px",
          }
        }
      />
    </a>
    <div
      style={
        Object {
          "color": "white",
          "left": "50%",
          "position": "absolute",
        }
      }
    >
      Annotation tool
    </div>
    <div
      style={
        Object {
          "position": "absolute",
          "right": "5px",
        }
      }
    >
      <div>
        <a
          download="annotations.json"
          href="data:text/json;charset=utf-8,%7B%22annotations%22%3A%5B%5D%2C%22relations%22%3A%5B%5D%2C%22text%22%3A%22Lorem%20ipsum%20dolor%20sit%20amet%2C%20nominavi%20rationibus%20vix%20in.%22%7D"
          style={
            Object {
              "color": "white",
              "padding": 5,
            }
          }
        >
          Download annotations
        </a>
        <span
          onClick={[Function]}
          style={
            Object {
              "color": "white",
              "cursor": "pointer",
              "padding": 5,
              "textDecoration": "underline",
            }
          }
        >
          New article
        </span>
      </div>
    </div>
  </div>,
  <div
    style={
      Object {
        "height": 60,
        "position": "relative",
        "width": "100%",
        "zIndex": 2,
      }
    }
  >
    <label
      style={
        Object {
          "paddingTop": 20,
          "position": "absolute",
          "right": "20px",
        }
      }
    >
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
      />
      Show relations
    </label>
  </div>,
  <div
    style={
      Object {
        "alignItems": "flex-start",
        "display": "flex",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <div
      id="article-text-content"
      style={
        Object {
          "paddingBottom": "70px",
        }
      }
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "height": "100%",
            "whiteSpace": "pre-wrap",
            "width": 900,
          }
        }
      >
        Lorem ipsum dolor sit amet, nominavi rationibus vix in.
      </div>
    </div>
  </div>,
]
`;
